### PR TITLE
Hierarchical GScan

### PR DESCRIPTION
**Pull request by kinow**
Related to #139 

* Brings hierarchy into GScan, using existing Tree, TreeItem, Job, and Task components.
* Creates modules for `sort`, `filter` (will spread this pattern to other components over next PR's/days/weeks/...).
* Makes the filter workflow name field case-insensitive (i.e. searching for `RUN1` will include a workflow like `five/run1`).
* Has a known bug in sorting by status, see comment below, and pending #703 
* I think I could try some performance improvements, for instance, for every delta received it goes over the tree, finds the next insertion index, filters, etc... if you have 40 workflows, you will receive 40 workflows. We could debounce it and run this every second or every 500ms, giving time for more deltas to be applied and reduce computation in JS and DOM updates (confirmed the DOM is actually updated while the deltas are applied if not super-fast :disappointed_relieved: ). But this is for a next PR, maybe :+1: 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

**Note for reviewers**: you may find workflows that are not sorted correctly by **status**. This is a known issue, with a fix already in review. This is what it may look like (`run4` is supposed to go up I think, as it is running, but Enumify is messing up).

image

p.s.: to test calling the callback for a random delta, execute the following in the browser console

```js
var deltas = [
{
  "workflow": {
    "id": "user|research/test/one/run2",
    "name": "research/test/one/run2",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "user|research/test/two/run1",
    "name": "research/test/two/run1",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "research",
    "name": "research",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
}
]
deltas.forEach((delta) => {
window.app.$store.dispatch('workflows/applyWorkflowsDeltas', { deltas: { added: delta } })
})
```

### CODE REVIEW

**Comment by kinow** [comment]
This new feature doesn't look too complicated to add, but I would like to see, first, if we can add it using the existing `Tree` and `TreeItem` components.

For that, we just need to make `Tree` and `TreeItem` as generic as possible. And also need a preprocessing step, to occur after the data is received, and before it is given to the `GScan` component, so that we have the hierarchy for workflows with slashes in their names.

**Comment by kinow** [comment]
Huh, not so straightforward :laughing: 

image

But so far less than 2 hours spent on this, so not too bad. Will continue tomorrow.

**Comment by kinow** [comment]
Getting there!

image

**Comment by kinow** [comment]
image

In the image above, the links to workflows are working. It is the UI with the latest commit here. It is missing to group hierarchies that have common parts (e.g. `a/1` and `a/b/c` share the `a` part, so they should have a common node starting the tree branch).

But **what happens when we have running workflows in these hierarchies**? At the moment we show running workflows at the top. But if we have a workflow like `a/1` running, and we moved the `a` node to the top, it could have `a/b/c` and several other workflows stopped under it too.

So would we still display that hierarchy at the top? Along with all other stopped workflows? And below it display other running workflows like `five`? Followed by stopped workflows?

**Comment by hjoliver** [comment]
Good question. We may need to be able to choose between a flat list with "global" sorting (of running/not-running) and a tree view with partial sorting (by running/not-running)

A flat list, particularly when combined with filtering, is easiest to use if you don't care about the hierarchical grouping.

A tree view can't be fully sorted like a flat list without duplication of branches, which I don't think we would want. 

I guess as a start, the tree view should sort (by running/not-running) in the leaves on each terminal branch.

And we could also consider sorting branches so that those with one or more running flows are at the top, and those with no running flows are at the bottom.

That's probably the best we can do, for sorting within a tree (is it?)?

**Comment by oliver-sanders** [comment]
Note we used to offer a flat option in the tree view for much the same reason (better re-implemented as a table if people *really* want that). So long as hierarchical is the default!

Note that hierarchical display makes more sense if we go down the hierarchical run registration route i.e:

```
~/
    cylc-run/
        flow-name/
            run1/
            run2/
```

Also, with this model the `flow-name` node could be used to represent the "template" suite providing a workspace where users could graph, validate and edit the Rose/Cylc configuration before running it.

**Comment by hjoliver** [comment]
@kinow I've had a quick play with this.  First thing that struck me was, the hierarchy is being displayed as separate hierarchicall items rather than as a unified tree form: 
foo

(The root and branches should merge going up the tree, right?)

**Comment by hjoliver** [comment]
I see you commented on that yourself above, but maybe my answer wasn't very helpful.  I was trying to say we can't have **both** a grouping tree **and** absolute sorting of leaves by running state. And I went off tangent a bit by saying we might also need a flat (non-tree) gscan with absolute sorting by state - but that's a potential addition for the future.  For now, we definitely want a tree gscan.  The gscan tree will have multiple trunks (i.e. multiple trees - like the multiple cycle point trunks in the tree view).  So we should sort trunks according to whether or not *any* of their leaf flows are running: those with any running flows at the top; those no running flows at the bottom.   And similarly inside the lower branches of any tree, down to the leaves (i.e. running leaves at the top of their own tree).  

Does that make sense?

**Comment by kinow** [comment]
>(The root and branches should merge going up the tree, right?)

Yup, I was working on that when I realized it would break sorting. 

From what I understood in the discussion here, it's OK to go with hierarchy, and sort by the complete name.

So if we have `a/b/c` and `aaa` and `c/b/a`, the order would simply be `aaa`, then `a/b/c` and then `c/b/a`. i.e. the sort compares the string `a/b/c`, not the workflow name (`c` in this case).

Also, filtering would be by workflow name, so it would use `a/b/c` and any string like "a", "b", "c", would not remove the workflow from the list.

And then later we can add a toggle to switch between flat and non-flat. Or would it be best to do that now? :thinking:

**Comment by hjoliver** [comment]
(Alphanumeric sorting might be wanted too, but that can be another PR).

**Comment by hjoliver** [comment]
> And then later we can add a toggle to switch between flat and non-flat. Or would it be best to do that now? 

A follow-up is fine unless it's really easy for you.

> From what I understood in the discussion here, it's OK to go with hierarchy, and sort by the complete name.

Well I was suggesting a combination of that and run/stopped sort.  

Trunks with *any running flows under them* should go above those with no running flows.  Within that, sort by name. And if you completely expand one trunk, running leaf-flows at the top and stopped ones below them - but sort all the running flows within that group by name, and all the stopped ones within that group by name.

**Comment by hjoliver** [comment]
(If that's still not very clear let's have a chat tomorrow - it's hard to explain this stuff in text! - I have to go out now to a school thing...)

**Comment by hjoliver** [comment]
I guess I'm just talking about primary and secondary sort keys.  Primary = running/stopped; secondary = alphanumeric. :boom: 
And that's for all the trees at the top level; and for intermediate branches within a tree; and for leaves within a tree.

**Comment by kinow** [comment]
Righto. I will try to implement as close as possible, but writing as little code as possible, so that and the others can take a look (some times looking at the running UI more ideas appear, or we change our mind), and also for me to take a look at performance. I remember having some issues when implementing hierarchy, and that was before the offline workflows I think? So now I will have even more workflows.

The first thing I will do is find a way to merge the families. So `a/1` and `a/2` both appear under a single `a` node :+1: then will start working on the sorting.

**Comment by hjoliver** [comment]
One thing that definitely won’t change is we don’t want to break the “families” up. Sorting, by whichever keys, can only change the order of the top level trees, and the order of nodes within expanded trees.

**Comment by kinow** [comment]
I am thinking in writing the code to handle the hierarchy, and try to use jsperf.com (which has been [502'ing](https://github.com/jsperf/jsperf.com/issues/537) for 1 month, but should be able to run locally) to confirm it will perform OK.

I am planning on leaving this PR open, while the other PR's are reviewed and merged, and work on 

- [ ] send a PR with the improvements for the Tree/TreeItem in a separate PR (which allows a more generic use for the tree and treeitem components)
- [ ] send another PR to use deltas in the GScan. I think this way I won't have to iterate the returned values to decide how to re-create or update the GScan hierarchy with every request.
- [ ] update this PR to use the code from the other two PR's, and then focus on only the hierarchy building

**Comment by kinow** [comment]
Whew rebased. Wrote down a few post-its & scribbles as I re-read the old code with some problems & improvements after the rebase. Will start working on these items tomorrow :neckbeard:

File: src/components/cylc/gscan/GScan.vue [hunk 1]
```diff
@@ -329,7 +329,7 @@ export default {
             { numeric: true, sensitivity: 'base' })
       })
         .map(workflow => {
-          return createWorkflowNode(workflow)
+          return createWorkflowHierarchyFromName(workflow)
```
Comment (kinow) [inline src/components/cylc/gscan/GScan.vue:332]: Note to self: maybe use just `createWorkflowNode`, but with a flag telling whether we want hierarchy or not? Or leave as-is…

File: src/components/cylc/tree/nodes.js [hunk 2]
```diff
@@ -343,11 +343,64 @@ function createJobNode (job) {
...
+      createWorkflowNode(workflow),
+      {
+        text: workflowPart,
+        children: null
+      }
+    )
+  )
+  return workflowHierarchy
+}
+
```
Comment (kinow) [inline src/components/cylc/tree/nodes.js:397]: Note to self: can we simplify this?

**Comment by kinow** [comment]
Just missing sorting now, then after that will take a quick break, code-review it again, and if I don't find anything wrong, start adding the tests and finally request the review :tada:  It may be ready by this week :crossed_fingers: 

Here's a GIF of what it's looking like so far.

GIFrecord_2021-07-19_122715

**Comment by kinow** [comment]
Sorting done :sleepy: :sweat_drops: 

image

The elements should be in order. Also used deltas out of order, to confirm it wasn't just accidentally adding elements in the right position :rofl: Expand code below for the deltas used to produce the screenshot above.

<details>
  <summary>Expand</summary>

```js
var deltas = [
{
  "workflow": {
    "id": "user|research/test/one/run2",
    "name": "research/test/one/run2",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "user|research/test/one/run0",
    "name": "research/test/one/run0",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "user|100",
    "name": "100",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "user|1",
    "name": "1",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "user|2",
    "name": "2",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "user|a",
    "name": "a",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "user|research/test/two/run1",
    "name": "research/test/two/run1",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
},
{
  "workflow": {
    "id": "research",
    "name": "research",
    "status": "running",
    "owner": "user",
    "host": "localhost",
    "port": 43079,
    "stateTotals": {
      "waiting": 1,
      "expired": 0,
      "preparing": 0,
      "submit-failed": 0,
      "submitted": 3,
      "running": 1,
      "failed": 0,
      "succeeded": 0
    },
    "latestStateTasks": {
      "running": [
        "running-task"
      ]
    }
  }
}
]
deltas.forEach((delta) => {
window.app.$store.dispatch('workflows/applyWorkflowsDeltas', { deltas: { added: delta } })
})
```

</details>

What's pending now:

- [ ] Review my own code before asking others to review it
- [ ] Add notes in the code for future reviewers
- [ ] Wait for #703 then rebase and fix conflicts
- [ ] Fix unit tests
- [ ] Fix e2e tests
- [ ] Wait for codecov or run coverage locally, see what was not tested and decide what more tests we need
- [ ] Write new unit tests
- [ ] Write e2e tests
- [ ] Changelog
- [ ] Request review :1st_place_medal:

File: src/services/mock/json/GscanSubscriptionQuery.json [hunk 3]
```diff
@@ -1,7 +1,7 @@
 {
   "workflow": {
-    "id": "user|one",
-    "name": "one",
+    "id": "user|research/test/one/run1",
+    "name": "research/test/one/run1",
```
Comment (kinow) [inline src/services/mock/json/GscanSubscriptionQuery.json:4]: Note to self: undo this change. It was only for development.

File: src/components/cylc/tree/sort.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+export function sortTaskProxyOrFamilyProxy (leftObject, leftValue, rightObject, rightValue) {
+  if (leftObject.type < rightObject.type) {
+    return -1
+  }
+  if (leftObject.type > rightObject.type) {
+    return 1
+  }
+  // name
+  return DEFAULT_COMPARATOR(leftValue, rightValue)
+}
```
Comment (kinow) [inline src/components/cylc/tree/sort.js:39]: I've created a `${component-name}/sort.js` for each component that sorts nodes. Now you can find the code to create nodes or node-types in `${component-name}/nodes.js`, filtering code in `${component-name}/filter.js`, general code in `${component-name}/index.js`, deltas code in `${component-name}/deltas.js`, etc.

That organizes the code in logical units, and reduces the amount of code in each file. Plus, once you learn how to work on one component, just expand the module/folder of another component and almost immediately you should know what you are looking at, and where are the things you need for your change.

File: src/components/cylc/tree/index.js [hunk 5]
```diff
@@ -61,116 +63,6 @@ function computeCyclePointsStates (cyclePointNodes) {
...
-    const higher = comparatorFunction(value, newValue, array[mid], midValue)
-    if (higher) {
-      low = mid + 1
-    } else {
-      high = mid
-    }
-  }
-  return high
-}
-
```
Comment (kinow) [inline src/components/cylc/tree/index.js:173]: Moved to a `sort.js` module.

File: src/components/cylc/gscan/sort.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+export function sortWorkflowNamePartNodeOrWorkflowNode (leftObject, leftValue, rightObject, rightValue) {
+  if (leftObject.node.type !== rightObject.node.type) {
+    return WORKFLOW_TYPES_ORDER.indexOf(leftObject.node.type) - WORKFLOW_TYPES_ORDER.indexOf(rightObject.node.type)
+  }
+  if (leftObject.node.status !== rightObject.node.status) {
+    const leftState = WorkflowState.enumValueOf(leftObject.node.status.toUpperCase())
+    const rightState = WorkflowState.enumValueOf(rightObject.node.status.toUpperCase())
+    // FIXME: waiting for https://github.com/cylc/cylc-ui/pull/703
+    //        which fixes another issue in GScan... Note to self,
+    //        look at the DIFF in GitHub when resolving conflicts!
```
Comment (kinow) [inline src/components/cylc/gscan/sort.js:51]: Note to self: remove me when the PR has been merged, and conflicts fixed... if the conflicts have been fixed and you are still seeing this code... well, you messed up when fixing conflicts.

File: src/components/cylc/gscan/sort.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * Sort workflows by type first, showing running or paused workflows first,
+ * then stopped. Within each group, workflows are sorted alphabetically
+ * (natural sort).
+ * @param leftObject
+ * @param leftValue
+ * @param rightObject
+ * @param rightValue
+ * @returns {boolean|number}
+ */
+export function sortWorkflowNamePartNodeOrWorkflowNode (leftObject, leftValue, rightObject, rightValue) {
```
Comment (kinow) [inline src/components/cylc/gscan/sort.js:42]: Used when adding new nodes to the GScan hierarchy/tree.

File: src/components/cylc/gscan/nodes.js [hunk 8]
```diff
@@ -15,27 +15,149 @@
...
+    // Here we calculate what is the index for this element. If we decide to have ASC and DESC,
+    // then we just need to invert the location of the element, something like
+    // `sortedIndex = (array.length - sortedIndex)`.
+    const sortedIndex = sortedIndexBy(
+      nodes,
+      node,
+      (n) => n.name,
+      sortWorkflowNamePartNodeOrWorkflowNode
+    )
+    nodes.splice(sortedIndex, 0, node)
```
Comment (kinow) [inline src/components/cylc/gscan/nodes.js:148]: Adding in order :point_up:

File: src/components/cylc/gscan/GScan.vue [hunk 9]
```diff
@@ -128,9 +134,9 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
                     />
                     <v-tooltip top>
                       <template v-slot:activator="{ on }">
-                        <span v-on="on">{{ scope.node.node.name || scope.node.id }}</span>
+                        <span v-on="on">{{ scope.node.name }}</span>
                       </template>
-                      <span>{{ scope.node.node.name || scope.node.id }}</span>
+                      <span>{{ scope.node.id }}</span>
```
Comment (kinow) [inline src/components/cylc/gscan/GScan.vue:139]: Display ID in tooltip.

File: src/components/cylc/gscan/GScan.vue [hunk 10]
```diff
@@ -128,9 +134,9 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
                     />
                     <v-tooltip top>
                       <template v-slot:activator="{ on }">
-                        <span v-on="on">{{ scope.node.node.name || scope.node.id }}</span>
+                        <span v-on="on">{{ scope.node.name }}</span>
```
Comment (kinow) [inline src/components/cylc/gscan/GScan.vue:137]: Display name in the tree. Note that this is not the GraphQL node name (i.e. `node.node.name`), but the `.name` property we have when creating the hierarchy.

For `folder/five/run1`, its name is `folder/five/run1`, but we have just `run1` to be used when displaying it. In the same manner, when we are adding the workflow-name-part `folder`, its `.name` will be... `folder` :+1: I accidentally modified the `node.node.name` changing its value, and when the deltas updated it in development I had a nasty bug :bug: took me near 1 hour to realize the mistake I had made :smile:

File: src/components/cylc/gscan/GScan.vue [hunk 11]
```diff
@@ -102,13 +102,13 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
       </div>
       <!-- data -->
       <div
-        v-if="!isLoading && sortedWorkflows && sortedWorkflows.length > 0"
```
Comment (kinow) [inline src/components/cylc/gscan/GScan.vue:105]: No more `sortedWorkflows` computed property. When GScan deltas are received, they are updated in Vuex via callbacks registered in the `WorkflowService`. When the Vuex state changes, it triggers a `computed` property that creates the hierarchical elements, adding elements in-order into the tree. Finally watchers will be notified and update the `filteredWorkflows` (i.e. apply filters, if any).

File: src/components/cylc/common/sort.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    const midValue = iterateeFunction(array[mid])
+    const higher = comparatorFunction(value, newValue, array[mid], midValue)
+    if (higher) {
+      low = mid + 1
+    } else {
+      high = mid
+    }
+  }
+  return high
+}
```
Comment (kinow) [inline src/components/cylc/common/sort.js:101]: Extracted from the Tree component, now used by the GScan component too.

File: src/services/mock/json/GscanSubscriptionQuery.json (hunk 3 above)
Comment (kinow) [reply src/services/mock/json/GscanSubscriptionQuery.json]: Undone!

**Comment by kinow** [comment]
Moved changelog to 0.6

**Review by hjoliver — APPROVED** [review]
Looks good, tests as working :+1: 

As a follow-up, we should have a summary state propagate up the tree (so you can see if anything is running in a collapsed branch...)

**Comment by kinow** [comment]
> As a follow-up, we should have a summary state propagate up the tree (so you can see if anything is running in a collapsed branch...)

https://github.com/cylc/cylc-ui/issues/721

**Review by oliver-sanders — APPROVED** [review]
Looks great, thanks!

One issue for follow-up, hopefully small although it might be difficult due to component structure? The hover/selection is indented and doesn't stretch right up to the right-hand margin. This can get a little messy for hierarchically named workflows.

Screenshot from 2021-08-12 13-45-12

**Comment by kinow** [comment]
> Looks great, thanks!
> 
> One issue for follow-up, hopefully small although it might be difficult due to component structure? The hover/selection is indented and doesn't stretch right up to the right-hand margin. This can get a little messy for hierarchically named workflows.
> 
> Screenshot from 2021-08-12 13-45-12

Ah, hadn't notice it! https://github.com/cylc/cylc-ui/issues/730 Thanks!
